### ISSUE TITLE
laparams=None is a TERRIBLE default!

### ISSUE THREAD

**Issue by snoopyjc** [pdfminer/pdfminer.six#247]
It took me hours with a debugger to figure out why the result of conversion to text was just running all the words together into one very long string!     Then I stumbled upon this comment: " laparams: An LAParams object from pdfminer.layout.
        Default is None but may not layout correctly."

This default should be changed to an empty LAParams so at least it lays out the text properly by default - or have None mean the same as an empty one.

**Comment by pietermarsman** [comment]
I agree with you that `laparams=None` does not give the best result and therefore is a terrible default.

However, changing it would be a  breaking change with earlier releases. I would suggest that we raise a warning if `laparams=None` is used, about that this is going to change in the future.

**Comment by snoopyjc** [comment]
Deprecate it! :-)

-joe

**Comment by pietermarsman** [comment]
If you like you can create a pull request that does that. Then we also have something concrete to talk about.

**Comment by jstockwin** [comment]
This issue has gone a bit stale. @snoopyjc Are you still interested in potentially working on this? @pietermarsman Do you know if it's still an issue at all?

Personally, I'd be up for changing the default. If `None` gives nonsense results, then nobody will be using it, so it's not really that much of a breaking change.

**Comment by pietermarsman** [comment]
Yes it is. I think it is most prone in `extract_text_to_fp`. 

- It uses `laparams=None` as a default. Other high level function use `laparams=LAParams()` if nothing is specified.
- It creates a `PDFLayoutAnalyzer` with the `None` laparams. In the `PDFLayoutAnalyzer.end_page()` method the layout analyzing is skipped if `self.laparams is None`.

Disabling layout analysis is a valid thing to do if someone only wants the pdf objects. However, it is indeed a bad default. Thats why other high level functions use `LAParams()` as a default.

**Comment by jstockwin** [comment]
So perhaps the existence of the high level functions closes this issue, and we should ensure the docs point to these as much as possible?

**Comment by snoopyjc** [comment]
I haven’t needed to use this module since last year and my stance that the default needs to change remains.

**Comment by jstockwin** [comment]
Hmm, if we don't want to change it because it's a breaking change, then perhaps we should highlight everywhere in the docs that you're likely to want to overwrite it?

**Comment by pietermarsman** [comment]
Maybe deprecate and remove `extract_text_to_fp` entirely? I've never liked it that you hand it an output stream and it writes the result to it. And we have better high-level function for extracting text and pages now. 

There are many examples of `extract_text_to_fp` on the internet so deprecating might take a while. But by deprecating it we can think of this issue being fixed. Because you shouldn't use the method at all.

We should probably add other methods to replace its functionality: `extract_html`, `extract_xml`, `extract_tags` and `extract_images`.

Another alternative is to raise a big warning if laparams=None is used.

**Comment by jstockwin** [comment]
Or both? I.e. deprecate the function but also add a warning? Agree we should add replacements first, though...

**Comment by pietermarsman** [comment]
So plan of attack: 

- Raise warnings when using `extract_text_to_fp`
    * A deprecation warning because the function will be removed in the future (when? 2021?)
    * A warning if `laparams=None` is used, that this is a terrible default and maybe a reference to this issue
- Replace the functionality of `extract_text_to_fp`
    * `extract_html`
    * `extract_xml`
    * `extract_tags`
    * `extract_images`
